src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -0,0 +1,185 @@
+    outColSize(0),
+    depth(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation(
+    const size_t inRowSize,
+    const size_t inColSize,
+    const size_t outRowSize,
+    const size_t outColSize,
+    const size_t depth):
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, 1);
+  else
+  {
+    assert(output.n_rows == outRowSize * outColSize * depth);
+    assert(output.n_cols == 1);
+  }
+
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  arma::cube inputAsCube(input.memptr(), inRowSize, inColSize, depth);
+  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize, depth,
+                          false, true);
+
+  double scaleRow = (double) inRowSize / (double) outRowSize;
+  double scaleCol = (double) inColSize / (double) outColSize;
+
+  arma::mat22 coeffs;
+  for (size_t i = 0; i < outRowSize; i++)
+  {
+    size_t rOrigin = (size_t) std::floor(i * scaleRow);
+    if (rOrigin > inRowSize - 2)
+      rOrigin = inRowSize - 2;
+
+    // Scaled distance of the interpolated point from the topmost row.
+    double deltaR = i * scaleRow - rOrigin;
+    if (deltaR > 1)
+      deltaR = 1.0;
+    for (size_t j = 0; j < outColSize; j++)
+    {
+      // Scaled distance of the interpolated point from the leftmost column.
+      size_t cOrigin = (size_t) std::floor(j * scaleCol);
+      if (cOrigin > inColSize - 2)
+        cOrigin = inColSize - 2;
+
+      double deltaC = j * scaleCol - cOrigin;
+      if (deltaC > 1)
+        deltaC = 1.0;
+      coeffs[0] = (1 - deltaR) * (1 - deltaC);
+      coeffs[1] = deltaR * (1 - deltaC);
+      coeffs[2] = (1 - deltaR) * deltaC;
+      coeffs[3] = deltaR * deltaC;
+
+      for (size_t k = 0; k < depth; k++)
+      {
+        outputAsCube(i, j, k) = arma::accu(inputAsCube.slice(k).submat(
+                rOrigin, cOrigin, rOrigin + 1, cOrigin + 1) % coeffs);
+      }
+    }
+  }
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /*input*/,
+    arma::Mat<eT>&& gradient,
+    arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(inRowSize * inColSize * depth, 1);
+  else
+  {
+    assert(output.n_rows == inRowSize * inColSize * depth);
+    assert(output.n_cols == 1);
+  }
+
+  assert(outRowSize >= 2);
+  assert(outColSize >= 2);
+
+  arma::cube gradientAsCube(gradient.memptr(), outRowSize, outColSize, depth);
```

Comment:
The same comment here applies about copying.  (You may have already realized it applies to both places but I want to explicitly point it out so we don't forget one.)